mobile_verifier/src/heartbeats.rs

```diff
@@ -94,167 +146,248 @@ impl HeartbeatDaemon {
+        &self,
+        reports: impl Stream<Item = Heartbeat> + 'a,
+        cache: &Cache<(String, DateTime<Utc>), ()>,
+        mut transaction: Transaction<'_, Postgres>,
+        epoch: &'a Range<DateTime<Utc>>,
+    ) -> anyhow::Result<()> {
+        let mut validated_heartbeats = pin!(
+            ValidatedHeartbeat::validate_heartbeats(&self.gateway_client, reports, epoch).await
+        );
 
-        while let Some(heartbeat) = validated_heartbeats.next().await.transpose()? {
-            heartbeat.write(&self.file_sink).await?;
+        while let Some(validated_heartbeat) = validated_heartbeats.next().await.transpose()? {
+            validated_heartbeat.write(&self.file_sink).await?;
 
-            if !heartbeat.is_valid() {
+            if !validated_heartbeat.check_validity() {
                 continue;
             }
 
-            let key = (heartbeat.cbsd_id.clone(), heartbeat.truncated_timestamp()?);
-
+            let key = validated_heartbeat.report.id()?;
             if cache.get(&key).await.is_none() {
-                heartbeat.save(&mut transaction).await?;
+                validated_heartbeat.save(&mut transaction).await?;
                 cache
                     .insert(key, (), time::Duration::from_secs(60 * 60 * 2))
                     .await;
             }
         }
-
         self.file_sink.commit().await?;
         transaction.commit().await?;
-
         Ok(())
     }
 }
 
-/// Minimum number of heartbeats required to give a reward to the hotspot.
-pub const MINIMUM_HEARTBEAT_COUNT: i64 = 12;
+#[derive(Debug, Clone, sqlx::FromRow)]
+pub struct HeartbeatRow {
+    pub hotspot_key: PublicKeyBinary,
+    // cell hb only
+    pub cbsd_id: Option<String>,
+    pub cell_type: CellType,
+    // wifi hb only
+    pub location_validation_timestamp: Option<DateTime<Utc>>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct HeartbeatReward {
+    pub hotspot_key: PublicKeyBinary,
+    pub cell_type: CellType,
+    // cell hb only
+    pub cbsd_id: Option<String>,
+    pub reward_weight: Decimal,
+}
+
+impl From<HeartbeatRow> for HeartbeatReward {
+    fn from(value: HeartbeatRow) -> Self {
+        Self {
+            hotspot_key: value.hotspot_key,
+            cell_type: value.cell_type,
+            cbsd_id: value.cbsd_id,
+            reward_weight: value.cell_type.reward_weight()
+                * value
+                    .cell_type
+                    .location_weight(value.location_validation_timestamp),
+        }
+    }
+}
 
 impl HeartbeatReward {
+    pub fn id(&self) -> anyhow::Result<String> {
+        match self.cell_type.to_label() {
+            CellTypeLabel::Cell => Ok(self
+                .cbsd_id
+                .clone()
+                .ok_or_else(|| anyhow!("expected cbsd_id, found none"))?),
+            CellTypeLabel::Wifi => Ok(self.hotspot_key.to_string()),
+            _ => Err(anyhow!("failed to derive label from cell type")),
+        }
+    }
+
     pub fn validated<'a>(
```

Comment:
i dunno, not sure i see much advantage from that approach other than avoiding a super large query ?  

Also having it as a single query does make it easier to test manually and confirm a uniform dataset is being returned

Reply:
You're going to be calling this function anyway, so I don't see any disadvantage for testing if you split it up into two smaller queries. This would also allow us to test each individually _if we wanted to_:

```rust
pub fn validated<'a>(pool: &'a PgPool, epoch: &'a Range<DateTime<Utc>>) -> Stream<Item = Result<HeartbeatReward>> {
    HeartbeatReward::validated_cbrs_heartbeats(pool, epoch).merge(HeartbeatReward::validated_wifi_heartbeats(pool, epoch))
}
```

Additionally, a cool thing with this is by using the [tokio merge StreamExt method](https://dtantsur.github.io/rust-openstack/tokio/stream/trait.StreamExt.html#method.merge), you can interleave both results as they are ready. That might end up being more efficient? I'm not a postgres expert so maybe not

Reply:
i am going to keep as is for time being.  I dont want to be messing with stream merges this close to going live.  Can revisit at a later point